File: e2e_tests/tools/custom/python/pekobot_adapter.py
#!/usr/bin/env python3
"""
Pekobot Universal Tool Adapter for Python (E2E Test Version)

Minimal adapter for E2E testing - handles JSON-RPC protocol over stdio.
"""

import sys
import json
import traceback
from dataclasses import dataclass
from typing import Any, Callable, Dict, List, Optional


@dataclass
class ExecutionContext:
    """Runtime context injected by Pekobot"""
    session_id: str
    agent_id: str
    workspace: str
    peer_id: Optional[str] = None
    run_id: Optional[str] = None


class Tool:
    """Decorator-based tool wrapper"""
    
    def __init__(
        self,
        name: str,
        description: str,
        parameters: Dict[str, Any] = None,
        reserved: List[str] = None,
        llm_description: str = None
    ):
        self.name = name
        self.description = description
        self.parameters = parameters or {}
        self.reserved = reserved or []
        self.llm_description = llm_description
        self.handler: Optional[Callable] = None
    
    def __call__(self, func: Callable) -> "Tool":
        """Use as decorator"""
        self.handler = func
        return self
    
    def run(self):
        """Start the protocol loop"""
        if not self.handler:
            raise RuntimeError("No handler registered")
        
        # Build parameter schema
        schema = {
            "type": "object",
            "properties": {},
            "required": []
        }
        
        for name, spec in self.parameters.items():
            schema["properties"][name] = spec
            if "default" not in spec:
                schema["required"].append(name)
        
        # Protocol loop
        for line in sys.stdin:
            line = line.strip()
            if not line:
                continue
            
            try:
                # Handle potential UTF-8 BOM
                line_clean = line.lstrip('\ufeff')
                request = json.loads(line_clean)
                response = self._handle_request(request, schema)
            except json.JSONDecodeError as e:
                response = self._error_response(None, -32700, f"Parse error: {e}")
            except Exception as e:
                response = self._error_response(None, -32603, f"Internal error: {e}")
            
            print(json.dumps(response), flush=True)
    
    def _handle_request(self, request: Dict[str, Any], schema: Dict[str, Any]) -> Dict[str, Any]:
        """Route request to appropriate handler"""
        method = request.get("method", "")
        req_id = request.get("id")
        
        if method == "tool/describe":
            return self._handle_describe(req_id, schema)
        elif method == "tool/execute":
            return self._handle_execute(req_id, request.get("params", {}))
        else:
            return self._error_response(req_id, -32601, f"Method '{method}' not found")
    
    def _handle_describe(self, req_id: Any, schema: Dict[str, Any]) -> Dict[str, Any]:
        """Return tool description"""
        result = {
            "name": self.name,
            "description": self.description,
            "parameters": schema
        }
        
        if self.llm_description:
            result["llm_description"] = self.llm_description
        
        if self.reserved:
            result["reserved_parameters"] = {
                name: {"source": "runtime", "description": f"Injected {name}"}
                for name in self.reserved
            }
        
        return self._success_response(req_id, result)
    
    def _handle_execute(self, req_id: Any, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute the tool handler"""
        try:
            args = params.get("args", {})
            context_data = params.get("context", {})
            
            # Merge reserved params into args
            merged_args = dict(args)
            
            # Inject reserved parameters from context
            for reserved in self.reserved:
                if reserved in context_data:
                    merged_args[reserved] = context_data[reserved]
            
            # Call handler
            result = self.handler(**merged_args)
            
            # Format result
            if result is None:
                result = {"success": True}
            elif not isinstance(result, dict):
                result = {"data": result}
            elif "success" not in result:
                result["success"] = True
            
            return self._success_response(req_id, result)
            
        except Exception as e:
            traceback.print_exc(file=sys.stderr)
            return self._success_response(req_id, {
                "success": False,
                "error": str(e)
            })
    
    def _success_response(self, req_id: Any, result: Any) -> Dict[str, Any]:
        return {
            "jsonrpc": "2.0",
            "id": req_id,
            "result": result
        }
    
    def _error_response(self, req_id: Any, code: int, message: str) -> Dict[str, Any]:
        return {
            "jsonrpc": "2.0",
            "id": req_id,
            "error": {
                "code": code,
                "message": message
            }
        }


def tool(
    name: str,
    description: str,
    parameters: Dict[str, Any] = None,
    reserved: List[str] = None,
    llm_description: str = None
) -> Tool:
    """Create a tool decorator"""
    return Tool(
        name=name,
        description=description,
        parameters=parameters,
        reserved=reserved,
        llm_description=llm_description
    )

File: e2e_tests/tools/custom/node/pekobot_adapter.js
#!/usr/bin/env node
/**
 * Pekobot Universal Tool Adapter for Node.js (E2E Test Version)
 * 
 * Minimal adapter for E2E testing - handles JSON-RPC protocol over stdio.
 */

const readline = require('readline');

/**
 * Create a tool definition
 * @param {Object} config - Tool configuration
 * @param {Function} handler - Tool handler function
 */
function tool(config, handler) {
    return {
        config: {
            name: config.name,
            description: config.description,
            parameters: config.parameters || {},
            reserved: config.reserved || [],
            llm_description: config.llm_description
        },
        handler
    };
}

/**
 * Run the protocol loop
 * @param {Object} toolDef - Tool definition from tool()
 */
function run(toolDef) {
    const schema = {
        type: "object",
        properties: {},
        required: []
    };

    // Build parameter schema
    for (const [name, spec] of Object.entries(toolDef.config.parameters)) {
        schema.properties[name] = spec;
        if (!spec.default !== undefined) {
            schema.required.push(name);
        }
    }

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: false
    });

    rl.on('line', (line) => {
        const trimmed = line.trim();
        if (!trimmed) return;

        try {
            const request = JSON.parse(trimmed);
            const response = handleRequest(request, toolDef, schema);
            console.log(JSON.stringify(response));
        } catch (e) {
            console.log(JSON.stringify(errorResponse(null, -32700, `Parse error: ${e.message}`)));
        }
    });
}

/**
 * Handle incoming request
 */
function handleRequest(request, toolDef, schema) {
    const { method, id, params = {} } = request;

    if (method === 'tool/describe') {
        return handleDescribe(id, toolDef, schema);
    } else if (method === 'tool/execute') {
        return handleExecute(id, params, toolDef);
    } else {
        return errorResponse(id, -32601, `Method '${method}' not found`);
    }
}

/**
 * Handle tool/describe
 */
function handleDescribe(id, toolDef, schema) {
    const result = {
        name: toolDef.config.name,
        description: toolDef.config.description,
        parameters: schema
    };

    if (toolDef.config.llm_description) {
        result.llm_description = toolDef.config.llm_description;
    }

    if (toolDef.config.reserved.length > 0) {
        result.reserved_parameters = {};
        for (const name of toolDef.config.reserved) {
            result.reserved_parameters[name] = {
                source: "runtime",
                description: `Injected ${name}`
            };
        }
    }

    return successResponse(id, result);
}

/**
 * Handle tool/execute
 */
async function handleExecute(id, params, toolDef) {
    try {
        const args = params.args || {};
        const context = params.context || {};

        // Merge reserved params from context
        const mergedArgs = { ...args };
        for (const reserved of toolDef.config.reserved) {
            if (context[reserved] !== undefined) {
                mergedArgs[reserved] = context[reserved];
            }
        }

        // Call handler
        const result = await toolDef.handler(mergedArgs);

        // Format result
        let formattedResult;
        if (result === null || result === undefined) {
            formattedResult = { success: true };
        } else if (typeof result !== 'object') {
            formattedResult = { success: true, data: result };
        } else {
            formattedResult = { success: true, ...result };
            if (!formattedResult.hasOwnProperty('success')) {
                formattedResult.success = true;
            }
        }

        return successResponse(id, formattedResult);
    } catch (e) {
        console.error(e);
        return successResponse(id, {
            success: false,
            error: e.message
        });
    }
}

/**
 * Create success response
 */
function successResponse(id, result) {
    return {
        jsonrpc: "2.0",
        id,
        result
    };
}

/**
 * Create error response
 */
function errorResponse(id, code, message) {
    return {
        jsonrpc: "2.0",
        id,
        error: { code, message }
    };
}

module.exports = { tool, run };

// If run directly, start demo echo tool
if (require.main === module) {
    const echoTool = tool(
        {
            name: "echo",
            description: "Echoes back the input",
            parameters: {
                message: { type: "string", description: "Message to echo" }
            }
        },
        ({ message }) => {
            return { echo: message };
        }
    );
    
    run(echoTool);
}
